src/utils.cpp

```diff
@@ -350,6 +356,216 @@ tagint utils::tnumeric(const char *file, int line, const char *str,
   return ATOTAGINT(str);
 }
 
+/* ----------------------------------------------------------------------
+   compute bounds implied by numeric str with a possible wildcard asterisk
+------------------------------------------------------------------------- */
+void utils::bounds(const char *file, int line, char *str,
+                   int nmin, int nmax, int &nlo, int &nhi, Error *error)
+{
+  char *ptr = strchr(str,'*');
+
+  nlo = nhi = -1;
+  if (ptr == NULL) {
+    nlo = nhi = atoi(str);
+  } else if (strlen(str) == 1) {
+    nlo = nmin;
+    nhi = nmax;
+  } else if (ptr == str) {
+    nlo = nmin;
+    nhi = atoi(ptr+1);
+  } else if (strlen(ptr+1) == 0) {
+    nlo = atoi(str);
+    nhi = nmax;
+  } else {
+    nlo = atoi(str);
+    nhi = atoi(ptr+1);
+  }
+
+  if (error) {
+    if (nlo < nmin)
+      error->all(file,line,fmt::format("Numeric index {} is out of bounds"
+                                       "({}-{})",nlo,nmin,nmax));
+    else if (nhi > nmax)
+      error->all(file,line,fmt::format("Numeric index {} is out of bounds"
+                                       "({}-{})",nhi,nmin,nmax));
+    else if (nlo > nhi)
+      error->all(file,line,fmt::format("Numeric index {} is out of bounds"
+                                       "({}-{})",nlo,nmin,nhi));
+  }
+}
+
+
+/* ----------------------------------------------------------------------
+   compute bounds implied by numeric str with a possible wildcard asterisk
+------------------------------------------------------------------------- */
+void utils::boundsbig(const char *file, int line, char *str,
```

Comment:
essentially the same function as `bounds()`, just with different type. `bounds()` could be a template, just needs a bit of metaprogramming for taking care of `atoi` vs. `ATOBIGINT`. I guess the preferences are to duplicate code vs having templates, correct?

Reply:
I've looked into making this a template function (having only `bounds()` and having it adapt would be nice), but I couldn't figure out a good way to make it so the compiler will fail to match it with the available signatures. The problem with  ATOBIGINT() could be handled with a specialization, but we would also have to force nmin and nmax to be `bigint`.
With the specialization we would have two copies of the function as well (unless you know a better way), and since there is only one use of boundsbig() I figured it is not worth the effort. If you can find an alternative, I would be willing to try.

Reply:
@rbberger did the change with promoting nmin/nmax to bigint and making the function signature a template function anyway, since this simplifies the use. The implementation remains as two specialization and no generic implementation.
I forgot about one more kink: when doing the specializations, one also has to consider the `bigint == int` case for `-DLAMMPS_SMALLSMALL`, so the second specialization must be ifdef'd out in that case. Will push this momentarily, just waiting for some final confirmation that it compiles with all "sizes".

Reply:
after some more thinking, it looks like I found a solution. in the process, I am also modernizing the code to use C++ strings and its associated methods.